src/three-components/CachingGLTFLoader.js

```diff
@@ -14,46 +14,71 @@
  */
 
 import GLTFLoader from '../third_party/three/GLTFLoader.js';
+import {cloneGltf} from './ModelUtils.js';
 
 const loadWithLoader = (url, loader) => new Promise((resolve, reject) => {
   loader.load(url, resolve, () => {}, reject);
 });
 
 const cache = new Map();
+const preloaded = new Map();
 
 export class CachingGLTFLoader {
+  static clearCache() {
+    cache.clear();
+    preloaded.clear();
+  }
+
   static has(url) {
     return cache.has(url);
   }
 
+  /**
+   * Returns true if the model that corresponds to the specified url is
+   * available in our local cache.
+   */
+  static hasFinishedLoading(url) {
+    return !!preloaded.get(url);
+  }
+
   constructor() {
     this.loader = new GLTFLoader();
   }
 
-  async load(url) {
+  /**
+   * Preloads a glTF, populating the cache. Returns a promise that resolves
+   * when the cache is populated.
+   */
+  async preload(url) {
     if (!cache.has(url)) {
-      cache.set(url, loadWithLoader(url, this.loader));
+      const loadAttempt = loadWithLoader(url, this.loader);
+      loadAttempt
+          .then(() => {
+            preloaded.set(url, true);
+          })
+          .catch(() => {});  // Silently ignore exceptions here, they should be
```

Comment:
Any reason using `.then` instead of await? This will swallow exceptions as indicated, but how would invoking function catch that if they're not thrown?

Reply:
When you invoke `then` or `catch` in a promise, it does not modify the state of a promise in place. The methods return _new_ promise instances that are resolved differently depending on how you invoked them.

We don't cache the "caught" promise. So, `try`/`catch` blocks outside of this scope will still work as expected.

Reply:
And yeah, the reason for using `then` is that the asynchronous behavior is preferred.

Reply:
As mentioned in https://github.com/GoogleWebComponents/model-viewer/pull/421#discussion_r265739701 I'll re-evaluate if this is necessary.

Reply:
Rewritten with `await`.